apps/labs/posts/python-wheels-from-tags-to-variants.md

```diff
@@ -0,0 +1,1083 @@
+  on x86-64 architecture (all that encoded in the `manylinux_2_34_x86_64`
+  platform tag). This is the wheel that will be used on most of the CPython
+  versions on the given platform.
+
+- `pp311-pypy311_pp73-manylinux_2_34_x86_64` designates a wheel built
+  for PyPy 3.11 (`pp311`) versions using `pypy311_pp73` ABI on the same
+  platform. Wheels like these are targeting specific Python interpreter
+  versions. They are used when the stable ABI is not used. Since some
+  implementations, such as PyPy or the freethreading CPython builds,
+  do not support the stable ABI, these are the wheels providing support
+  for them.
+
+- `py3-none-any` designates a wheel compatible with any Python 3.x
+  version, on any platform. It means a pure Python wheel, hence for pure Python
+  packages this is the only wheel needed. It can also be provided for packages
+  with optional extension modules,
+  as a fallback when there is no other wheel compatible with the user's system.
+
+You can find some unorthodox uses too, such as:
+
+- A pure Python package including dedicated code for every Python
+  version, for example using `py312-none-any` to designate a wheel
+  dedicated to Python 3.12 (or newer — though in this case newer Python
+  versions have their own wheels).
+
+- A pure executable package (that does not link to the Python library),
+  for example using `py3-none-macosx_11_0_arm64` to designate a wheel
+  compatible with any Python 3 implementation on a macOS 11.0+ system
+  with ARM64 CPU.
+
+Tags clearly provide some degree of flexibility. The existing code can usually
+account for future Python versions, operating system versions, libc versions,
+and to some degree for new implementations, ABI changes, new architectures
+and so on. But can it manage entirely novel tags?
+
+Well, technically yes. Some use cases would fit well within the current platform
+tags. Say, we could replace the plain `manylinux_2_28_x86_64` tag with a more
+fine-grained `manylinux_2_28_x86_64_v3`. Properties that aren't exclusive
+to existing platforms could be appended to them — say,
+`manylinux_2_28_x86_64_cuda129`, `manylinux_2_28_aarch64_openblas_openmpi`
+and so on. However, there are two major problems with that.
+
+First, the existing [pip](https://pypi.org/project/pip/) code, based
+on the [packaging](https://pypi.org/project/packaging/) library, relies entirely
+on enumerating all tag combinations compatible with the specific system
+and comparing the tags found in wheels to that list. For example,
+on a GNU/Linux system with Python 3.14 and glibc 2.41, this means almost 1300
+combinations. While this is not that much, once we start adding multiple
+additional suffixes, the number of possible combinations is going to explode.
+Of course, this isn't really a blocker — it is entirely possible to rewrite
+the logic to be more algorithmic and avoid generating all possible combinations.
+However, this is infeasible in practice.
```

Comment:
It's a bit weird to say it's not really a blocker and then say it's infeasible in practice. This doesn't feel super compelling as written. I might just drop it? If you keep it, I'd maybe say...

> While it's possible to optimize the algorithms that checks for tag compatibility, this becomes is infeasible in practice.